File: src/GlassCatalogIo.cpp
// this file is covered by the General Public License version 2 or later
// please see GPL.html for more details and licensing issues
// copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

#include <sstream>
#include <fstream>
using namespace std;

#include "GlassCatalogIo.h"
#include "GlassSellmeier.h"
#include "GlassExtended2.h"
#include "GlassManager.h"

////////////////////////////////////////////////////////////////////////////////
bool GlassCatalogIO::load(string sFile,GlassManager& pManager)
{
    ifstream f(sFile.c_str(),ios::in);

    bool bGlassPending=false;

    string sName;
    int iGlassFormula = 0;
    string sMaker;
    double B1=0.;
    double B2=0.;
    double B3=0.;
    double C1=0.;
    double C2=0.;
    double C3=0.;
    double B4=0.;
    double B5=0.;

    auto iIndexStartMaker=sFile.find_last_of('\\');
    if( (iIndexStartMaker!=string::npos) )
        sMaker=sFile.substr(iIndexStartMaker+1);
    auto iIndexSeparator=sMaker.find_first_of("_. -");
    if( (iIndexSeparator!=string::npos) )
        sMaker=sMaker.substr(0,iIndexSeparator);

    while(!f.eof())
    {
        string sLine;
        std::getline(f,sLine);

        //trim first space if any
        size_t iPos=sLine.find_first_not_of(" ");
        if(iPos!=string::npos)
            sLine=sLine.substr(iPos);

        if(sLine.size()<2)
            continue; // nothing to parse //TODO test bGlassPending before exit

        string sKey=sLine.substr(0,2);
        string sVal=sLine.substr(2); // can be empty

        //trim first space if any
        size_t iPos2=sVal.find_first_not_of(" ");
        if(iPos2!=string::npos)
            sVal=sVal.substr(iPos2);

        if(sKey=="CC")
        {
            //	pCatalog->set_comment(sVal);
        }

        if(sKey=="NM")
        {
            if(bGlassPending)
            {
                if(iGlassFormula==2) //Sellmeier
                {
                    //create and store the glass
                    GlassSellmeier* pGlass=new GlassSellmeier;
                    pGlass->set_maker(sMaker);
                    pGlass->set_name(sName);
                    pGlass->set_coefs(B1,B2,B3,C1,C2,C3);
                    pGlass->set_solid_color(8844025); //TODO
                    pManager.inject(pGlass);
                }
                else if(iGlassFormula==12) //Extended2
                {
                    //create and store the glass
                    GlassExtended2* pGlass=new GlassExtended2;
                    pGlass->set_maker(sMaker);
                    pGlass->set_name(sName);
                    pGlass->set_coefs(B1,C1,B2,C2,B3,C3,B4,B5);
                    pGlass->set_solid_color(8844025); //TODO
                    pManager.inject(pGlass);
                }
                else
                {
                    //TODO
                }
            }

            //compute subkey
            stringstream ss(sVal);
            ss >> sName >> iGlassFormula;

            bGlassPending=true;
        }

        if(sKey=="CD") //GlassCoef formula
        {
            stringstream ss(sVal);
            ss >> B1 >> C1 >> B2 >> C2 >> B3 >> C3 >> B4 >> B5; //may have more coef than needed
        }

        //	pCatalog->set_maker("unknown"); //TODO
    }

    return true;
}
////////////////////////////////////////////////////////////////////////////////

File: python/GlassCatalogIo.py
# this file is covered by the  GNU LESSER GENERAL PUBLIC LICENSE Version 3 or later
# please see LICENSE.txt for more details and licensing issues
# copyright Etienne de Foras ( the author )  mailto: etienne.deforas@gmail.com

import os

import GlassSellmeier
import GlassExtended2

class GlassCatalogIO:
    def load(self, file_path, glass_manager):
        with open(file_path, 'r') as file:
            glass_pending = False

            glass_name = ""
            glass_formula = 0
            glass_maker = ""
            coefficient_B1 = 0.0
            coefficient_B2 = 0.0
            coefficient_B3 = 0.0
            coefficient_C1 = 0.0
            coefficient_C2 = 0.0
            coefficient_C3 = 0.0
            coefficient_B4 = 0.0
            coefficient_B5 = 0.0

            index_start_maker = file_path.rfind('\\')
            if index_start_maker != -1:
                glass_maker = file_path[index_start_maker + 1:]
            index_separator = next((i for i, char in enumerate(glass_maker) if char in "_. -"), None)
            if index_separator is not None:
                glass_maker = glass_maker[:index_separator]

            for line in file:
                line = line.lstrip()  # trim first space if any

                if len(line) < 2:
                    continue  # nothing to parse

                key = line[:2]
                value = line[2:].lstrip()  # can be empty

                if key == "CC":
                    # pCatalog.set_comment(value)
                    pass

                if key == "NM":
                    if glass_pending:
                        if glass_formula == 2:  # Sellmeier
                            # create and store the glass
                            glass_sellmeier = GlassSellmeier()
                            glass_sellmeier.set_maker(glass_maker)
                            glass_sellmeier.set_name(glass_name)
                            glass_sellmeier.set_coefs(coefficient_B1, coefficient_B2, coefficient_B3, coefficient_C1, coefficient_C2, coefficient_C3)
                            glass_manager.inject(glass_sellmeier)
                        elif glass_formula == 12:  # Extended2
                            # create and store the glass
                            glass_extended2 = GlassExtended2()
                            glass_extended2.set_maker(glass_maker)
                            glass_extended2.set_name(glass_name)
                            glass_extended2.set_coefs(coefficient_B1, coefficient_C1, coefficient_B2, coefficient_C2, coefficient_B3, coefficient_C3, coefficient_B4, coefficient_B5)
                            glass_manager.inject(glass_extended2)
                        else:
                            # TODO
                            pass

                    # compute subkey
                    values = value.split()
                    glass_name = values[0]
                    glass_formula = int(values[1]) if len(values) > 1 else 0

                    glass_pending = True

                if key == "CD":  # GlassCoef formula
                    coefficients = list(map(float, value.split()))
                    coefficient_B1, coefficient_C1, coefficient_B2, coefficient_C2, coefficient_B3, coefficient_C3, coefficient_B4, coefficient_B5 = coefficients

        return True